sycl/include/sycl/ext/intel/experimental/esimd/detail/elem_type_traits.hpp

```diff
@@ -0,0 +1,568 @@
+//==------------ - elem_type_traits.hpp - DPC++ Explicit SIMD API ----------==//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+// This header provides:
+// - meta interfaces ("traits") which must be implemented to supoprt
+//   non-standard element types, such as sycl::half or sycl::bfloat16
+// - interfaces for performing various C++ operations on the types (+, *,...)
+//   and their default implementations.
+// - interfaces for performing conversions to/from the types and their default
+//   implementations.
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/ext/intel/experimental/esimd/detail/types.hpp>
+
+#include <CL/sycl/half_type.hpp>
+
+__SYCL_INLINE_NAMESPACE(cl) {
+namespace sycl {
+namespace ext {
+namespace intel {
+namespace experimental {
+namespace esimd {
+namespace detail {
+
+enum class BinOp {
+  ARITH_FIRST,
+  add = ARITH_FIRST,
+  sub,
+  mul,
+  div,
+  rem,
+  ARITH_LAST = rem,
+  BIT_FIRST,
+  shl = BIT_FIRST,
+  shr,
+  BIT_LOG,
+  bit_or = BIT_LOG,
+  bit_and,
+  bit_xor,
+  BIT_LST = bit_xor,
+  LOG_FIRST,
+  log_or = LOG_FIRST,
+  log_and,
+  LOG_LAST = log_and
+};
+
+enum class CmpOp {
+  CMP_FIRST,
+  lt = CMP_FIRST,
+  lte,
+  gte,
+  gt,
+  EQ_CMP_FIRST,
+  eq = EQ_CMP_FIRST,
+  ne,
+  CMP_LAST = ne
+};
+
+// If this element type is a special "wrapper" type. Such wrapper types, e.g.
+// sycl::half, bfloat etc, are used to represent non-standard element types in
+// user code.
+template <class T>
+static inline constexpr bool is_wrapper_elem_type_v =
+    std::is_same_v<T, sycl::half>;
+
+template <class T>
+static inline constexpr bool is_valid_simd_elem_type_v =
+    (is_vectorizable_v<T> || is_wrapper_elem_type_v<T>);
+
+struct invalid_storage_element_type;
+
+template <class T, class SFINAE> struct element_type_traits {
+  // The raw element type of the underlying clang vector used as a
+  // storage.
+  using StorageT = invalid_storage_element_type;
```

Comment:
Can probably rename StorageT to shorter RawT as you have used word 'raw' in many other places to describe storage type.

Reply:
ok